src/pnpq/devices/switch_thorlabs_osw_e.py

```diff
@@ -213,32 +224,103 @@ def get_state(self) -> State:
             return self._get_state()
 
     def _get_state(self) -> State:
-        """Private method to get the status of the switch without locks. This is used to check the status during set_state."""
+        """Private method to get the status of the switch without locks. This is used to check the status during set_state.
+
+        :raises ThorlabsOswError: If the device reports an error or sends an unexpected reply that cannot be parsed as a valid state.
+        """
         command = b"S?\n"
         self._connection.write(command)
         response = self._read_serial_response()
-        return State(int(response.decode("utf-8")))
+        try:
+            value_str = response.decode("utf-8")
+            value_int = int(value_str)
+            return State(value_int)
+        except Exception as e:
+            # This help to catch decoding erros and integer issues
+            raise ThorlabsOswError(
+                code=None,
+                description="Received invalid state response from Thorlabs OSW device.",
+                raw_reply=response.decode("utf-8", errors="replace"),
+            ) from e
 
     def get_type_code(self) -> str:
+        """
+        Get the board type code according to the configuration table.
+        :return: The type code as a human-readable, unstructured string.
+        :raises: ThorlabsOswError: If the device reports an error or sends an unexpected reply.
+        """
         self._fail_if_closed()
         with self._communication_lock:
             command = b"T?\n"
             self._connection.write(command)
             response = self._read_serial_response()
-            return response.decode("utf-8")
+            try:
+                return response.decode("utf-8")
+            except UnicodeDecodeError as e:
+                raise ThorlabsOswError(
+                    code=None,
+                    description="Received non-decodable type code from Thorlabs OSW device.",
+                    raw_reply=response.decode("utf-8", errors="replace"),
+                ) from e
```

Comment:
If you're worried about decoding errors, there's a better way to handle them: don't use `utf-8`. We could think of that as a bug in the application. Use a single-byte encoding scheme instead. Specifically, use `iso-8859-1`.

https://docs.python.org/3/library/codecs.html#module-encodings

Using utf-8 here was unlikely to fail (I'm not expecting characters outside of the basic 128 ASCII values) but using a single-byte encoding will remove the possibility of failure entirely, I think.

Also, dealing with this kind of error was not part of the assigned issue.